rclcpp/include/rclcpp/intra_process_manager_impl.hpp

```diff
@@ -184,6 +186,7 @@ class IntraProcessManagerState : public IntraProcessManagerStateBase
     size_t & size
   )
   {
+    std::lock_guard<std::mutex> lock(runtime_mutex_);
```

Comment:
Can this lock be moved into the next block, where we iterate over `publishers_`?  Or do we need to hold it all the way through the other two blocks?